File: Groupe2/TP2/src/couleurs.c
#include <stdio.h>
#include <stdbool.h>

#define MAX_COLORS 100

// Définition de la structure Color (typedef pour éviter de répéter `struct` à chaque fois)
typedef struct {
    unsigned char r, g, b, a;
} Color;

// Structure pour stocker les couleurs distinctes et leur nombre d'occurrences
typedef struct {
    Color color;
    int count;
} ColorCount;

// Fonction pour comparer deux couleurs (ici j'utilise bool de stdbool.h pour retourner vrai/faux, on pourrait aussi utiliser int pour retourner 1/0)
bool compare_colors(Color const c1, Color const c2) {
    return c1.r == c2.r && c1.g == c2.g && c1.b == c2.b && c1.a == c2.a;
}

// Fonction pour afficher une couleur
void print_color(Color const color) {
    printf("R: %d, G: %d, B: %d, A: %d : \033[48;2;%d;%d;%dm  \033[0m ", color.r, color.g, color.b, color.a, color.r, color.g, color.b);
}

int main() {
    // Tableau de couleurs (100 couleurs max)
    Color colors[MAX_COLORS] = {
        {255, 0, 0, 255}, {0, 255, 0, 255}, {0, 0, 255, 255},
        {255, 0, 0, 255}, {255, 0, 0, 255}, {0, 255, 0, 255},
        {238, 123, 241, 255}, {238, 123, 241, 255}, {238, 123, 241, 255}
    };

    // Tableau pour stocker les couleurs distinctes et leurs occurrences
    ColorCount distinct_colors[MAX_COLORS];
    int distinct_count = 0;

    // Parcourir toutes les couleurs
    for (int i = 0; i < MAX_COLORS; i++) {
        bool found = false;

        // Vérifier si la couleur est déjà dans distinct_colors
        for (int j = 0; j < distinct_count; j++) {
            if (compare_colors(colors[i], distinct_colors[j].color)) {
                distinct_colors[j].count++;
                found = true;
                break;
            }
        }

        // Si la couleur est nouvelle, l'ajouter au tableau distinct_colors
        if (!found) {
            distinct_colors[distinct_count].color = colors[i];
            distinct_colors[distinct_count].count = 1;
            distinct_count++;
        }
    }

    // Afficher les résultats
    printf("Couleurs distinctes et occurrences :\n");
    for (int i = 0; i < distinct_count; i++) {
        print_color(distinct_colors[i].color);
        printf(" => Occurrences: %d\n", distinct_colors[i].count);
    }

    return 0;
}

File: Groupe2/TP2/src/couleurs.py
class Color:
    def __init__(self, r, g, b, a):
        self.r = r
        self.g = g
        self.b = b
        self.a = a

    # __eq__ est une fonction par défaut dans une classe python qui permet de comparer deux objets
    def __eq__(self, other):
        return (self.r == other.r and
                self.g == other.g and
                self.b == other.b and
                self.a == other.a)

    # Méthode pour afficher une couleur
    def __str__(self):
        return f"R: {self.r}, G: {self.g}, B: {self.b}, A: {self.a}"


class ColorCount:
    def __init__(self, color):
        self.color = color
        self.count = 1


def print_color(color):
    # Fonction pour afficher une couleur et simuler une couleur de fond dans le terminal
    print(f"R: {color.r}, G: {color.g}, B: {color.b}, A: {color.a} ", end="")
    print(f"\033[48;2;{color.r};{color.g};{color.b}m  \033[0m", end="")


if __name__ == "__main__":
    # Tableau de couleurs prédéfinies (en python, il n'y a pas vraiment de limite contrairement en C)
    colors = [
        Color(255, 0, 0, 255), Color(0, 255, 0, 255), Color(0, 0, 255, 255),
        Color(255, 0, 0, 255), Color(255, 0, 0, 255), Color(0, 255, 0, 255),
        Color(238, 123, 241, 255), Color(238, 123, 241, 255), Color(238, 123, 241, 255)
    ]

    # Liste pour stocker les couleurs distinctes et leurs occurrences
    distinct_colors = []

    # Parcourir toutes les couleurs
    for color in colors:
        found = False  # bool est déjà un type primitif en python contrairement à C ou j'ai dû importer stdbool.h

        # Vérifier si la couleur est déjà dans distinct_colors
        for distinct_color in distinct_colors:
            if color == distinct_color.color:
                distinct_color.count += 1
                found = True
                break

        # Si la couleur est nouvelle, l'ajouter à la liste distinct_colors
        if not found:
            distinct_colors.append(ColorCount(color))

    # Afficher les résultats
    print("Couleurs distinctes et occurrences :")
    for distinct_color in distinct_colors:
        print_color(distinct_color.color)
        print(f" => Occurrences: {distinct_color.count}")
